opm/common/data/SimulationDataContainer.cpp

```diff
@@ -75,6 +75,36 @@ namespace Opm {
     }
 
 
+    void SimulationDataContainer::setCellDataComponent( const std::string& key , size_t component , const std::vector<int>& cells , const std::vector<double>& values) {
+        auto& data = getCellData( key );
+        if (component >= m_num_phases)
+            throw std::invalid_argument("Invalid component");
+
+        if (cells.size() != values.size())
+            throw std::invalid_argument("size mismatch between cells and values");
+
+        /* This is currently quite broken; the setCellDataComponent
+           method assumes that the number of components in the field
+           we are currently focusing on has num_phases components in
+           total. This restriction should be lifted by allowing a per
+           field number of components.
+        */
+        if (data.size() != m_num_phases * m_num_cells)
+            throw std::invalid_argument("Can currently only be used on fields with num_components == num_phases (i.e. saturation...) ");
+
+        for (size_t i = 0; i < cells.size(); i++) {
+            if (cells[i] < m_num_cells) {
+                auto field_index = cells[i] * m_num_phases + component;
+                auto value = values[i];
+
+                data[field_index] = value;
+            } else {
+                throw std::invalid_argument("Invalid cell number");
```

Comment:
Again, with OPM_THROW, the invalid cell number can easily be made part of the message.